### PR TITLE
Refacto kakarot.deploy and use .sol file as input for integration tests

### PR DESCRIPTION
**Pull request by ClementWalter**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

The `test_cases/erc20.py` uses a hard coded bytecode that is hard to maintain or verify.

## What is the new behavior?

The test runner first compile the original ERC20.sol file.
The output bytecode is then used in the subsequent tests.

## Other information

For now, I've added Solmate's ERC20 & ERC721.
I've slightly updated the contracts to make them deployable.

### CODE REVIEW

File: tests/solidity_files/output/ERC721TokenReceiver.abi [hunk 1]
```diff
@@ -0,0 +1 @@
+[{"inputs":[{"internalType":"address","name":"","type":"address"},{"internalType":"address","name":"","type":"address"},{"internalType":"uint256","name":"","type":"uint256"},{"internalType":"bytes","name":"","type":"bytes"}],"name":"onERC721Received","outputs":[{"internalType":"bytes4","name":"","type":"bytes4"}],"stateMutability":"nonpayable","type":"function"}]
```
Comment (Eikix) [inline tests/solidity_files/output/ERC721TokenReceiver.abi:1]: Why is this file compiled and not ERC721.bin / ERC721.abi ?

Comment (ClementWalter) [reply tests/solidity_files/output/ERC721TokenReceiver.abi:1]: ? the ERC721.sol indeed defined the ERC721 and the ERC721TokenReceiver, this makes a total of 3 contracts compiles, as found in the `output` folder

Comment (ClementWalter) [reply tests/solidity_files/output/ERC721TokenReceiver.abi:1]: but maybe we should just remove the `.abi` output as we only plan to use the `.bin`. The `*.opcode` may still be useful for debugging purposes

Comment (Eikix) [reply tests/solidity_files/output/ERC721TokenReceiver.abi:1]: yes but it says ERC721.bin is empty on my screen

image

**Comment by Eikix** [comment]
Tests break

**Review by Eikix — APPROVED** [review]
lgtm

**Comment by ClementWalter** [comment]
> Tests break

because it's not possible to actually deploy these contracts as is in kakarot, we need to either remove some part or to optimize the kakarot deployment. I just want to try something before cutting the .sol files

**Review by Eikix — CHANGES_REQUESTED** [review]
erc721.bin is still empty, is that normal / wanted ?

File: Makefile [hunk 2]
```diff
@@ -63,3 +63,6 @@ format-mac:
 	cairo-format src/**/*.cairo -i
 	black tests/.
 	isort tests/.
+
+build-sol:
+	docker run -v $(solidity_files):/sources ethereum/solc:stable -o /sources/output --abi --bin --overwrite --opcodes /sources/ERC20.sol /sources/ERC721.sol
```
Comment (matthieuauger) [inline Makefile:68]: Can we add docker in the prerequisites list in installation README?

Comment (ClementWalter) [reply Makefile:68]: good point

**Comment by ClementWalter** [comment]
> erc721.bin is still empty, is that normal / wanted ?

I'll update the base ERC721 file which is abstract to make it a deployable contract

File: src/kakarot/library.cairo [hunk 3]
```diff
@@ -170,45 +168,102 @@ namespace Kakarot {
...
         salt.write(value=current_salt + 1);
+
         // Generate EVM_contract address from the new cairo contract
         // TODO: TEMPORARY SOLUTION FOR HACK-LISBON !!!
-        let (_, low) = split_felt(contract_address);
-        local mock_evm_address = 0xAbdE100700000000000000000000000000000000 + low;
+        let (_, low) = split_felt(starknet_contract_address);
+        local evm_contract_address = 0xAbdE100700000000000000000000000000000000 + low;
+
+        evm_contract_deployed.emit(
```
Comment (Eikix) [inline src/kakarot/library.cairo:212]: nice:)

**Review by Eikix — APPROVED** [review]
lgtm

**Pull request merged**